### DISCUSSION TITLE
Filter does not work with upsert

### DISCUSSION THREAD

**Discussion by DecathectZero** [mikro-orm/mikro-orm#4829] [General]
**Describe the bug**
Postgres is being used with a soft deleted `deletedAt` column.

The `Filter` decorator [here](https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/decorators/Filter.ts) does not work with the `upsert` or `onConflict` commands. This applies to postgres and other SQL engines.

We have an entity with a soft delete column denoted by `deletedAt: Date`.

The following decorator is applying:
`@Filter({ name: 'soft_delete', cond: { deletedAt: { $eq: null } }, default: true })`

If we look a the raw query, this is what is being generated from this command `em.upsert(entity)`:

```
insert into "table" ("column") values ("value1") on conflict ("id") do update set "column" = excluded."column" returning "id", "deleted_at";
```

The correct query is:
```
insert into "table" ("column") values ("value1") on conflict ("id") where deleted_at IS NULL do update set "column" = excluded."column" returning "id", "deleted_at";
```

the `where deleted_at IS NULL` MUST follow up the conflict to be properly filtered.

**To Reproduce**
1. add a filter decorator to an entity, for example `Filter({ name: 'soft_delete', cond: { deletedAt: { $eq: null } }, default: true })`
2. call `upsert` on the entity
3. If there is a unique index on something + `deletedAt IS NULL`, SQL will reject the query entirely: `Query 1 ERROR: ERROR:  there is no unique or exclusion constraint matching the ON CONFLICT specification`

**Expected behavior**
the `where deleted_at IS NULL` MUST follow up the conflict to be properly filtered.
**Versions**

The code [here](https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/AbstractSqlDriver.ts#L461-L463
):
```
        qb.insert(data as T)
          .onConflict(uniqueFields)
          .returning(returning);
```
the `onConflict` must take in a "where" clause based on filter.

It should be closer to this:
```
        .insert(...)
        .onConflict(...)
        .where(filters)
        .merge()
        .returning(...)
```
Except the `onConflict` query always MOVES the `where` to after `merge` - which is incorrect behavior
This might be deeper and be an actual knex issue.

| Dependency | Version |
| - | - |
| node | 16^ |
| mikro-orm | 5.8.3 |

**Comment by B4nan** [comment]
Why should a filter be applied to upsert queries? Those are distinct features, filters are for reading, upsert is for insert/update.

> This might be deeper and be an actual knex issue.

Yes, I don't think what you want is even possible with knex.
